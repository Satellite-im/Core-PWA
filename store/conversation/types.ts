export enum ConversationConnection {
  CONNECTED = 'CONNECTED',
  CONNECTING = 'CONNECTING',
  DISCONNECTED = 'DISCONNECTED',
  DISCONNECTING = 'DISCONNECTING',
  ERROR = 'ERROR',
}

export enum ConversationActivity {
  ACTIVE = 'ACTIVE',
  INACTIVE = 'INACTIVE',
  TYPING = 'TYPING',
  NOT_TYPING = 'NOT_TYPING',
}

export type ConversationParticipant = {
  peerId?: string
  address: string
  name?: string // will be undefined for self
  profilePicture?: string
  state?: ConversationConnection
  activity?: ConversationActivity
  updatedAt?: number
  textilePubkey?: string // will be undefined for self
}

export type ConversationState = {
  id: string
  type: 'friend' | 'group'
  calling: boolean
  participants: Array<ConversationParticipant>
}
