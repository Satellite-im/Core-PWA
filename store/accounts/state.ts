import { AccountsState } from './types'

const InitalAccountsState: AccountsState = {
  locked: true,
  error: '',
  pinHash: '',
  active: '',
  gasPrice: '',
  phrase: '',
  encryptedPhrase: '',
}

export default InitalAccountsState
