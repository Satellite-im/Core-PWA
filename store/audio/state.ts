interface AudioState {
  muted: Boolean
  deafened: Boolean
  volume: Number
  inputVolume: Number
  previousVolume: Number
  sounds: {
    inboundMedia: Number
    outboundMedia: Number
    system: Number
  }
}

const InitalAudioState = (): AudioState => ({
  muted: false,
  deafened: false,
  volume: 100,
  inputVolume: 100,
  previousVolume: 0,
  sounds: {
    inboundMedia: 100,
    outboundMedia: 100,
    system: 100,
  },
})

export default InitalAudioState
