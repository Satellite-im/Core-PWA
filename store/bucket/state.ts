import { BucketState } from './types'

const InitialBucketState = (): BucketState => ({
  initialized: false,
  fileSystem: null,
})

export default InitialBucketState
