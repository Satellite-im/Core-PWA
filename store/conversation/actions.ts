import { PublicKey } from '@solana/web3.js'
import { keys } from 'libp2p-crypto'
import { createFromPubKey } from 'peer-id'
import {
  ConversationActivity,
  ConversationParticipant,
  ConversationState,
} from './types'
import { ActionsArguments } from '~/types/store/store'

const actions = {
  /**
   * @method initialize
   * @description Initialize the conversation store
   * @example
   * store.dispatch('conversation/initialize');
   */
  initialize({ commit }: ActionsArguments<ConversationState>) {
    commit('setConversation', {
      id: '',
      type: 'friend',
      calling: false,
      participants: [],
    })
  },
  /**
   * @method setConversation
   * @description Set the current conversation
   * @example
   * store.dispatch('conversation/setConversation', conversation);
   */
  setConversation(
    { commit }: ActionsArguments<ConversationState>,
    payload: {
      id: string
      type: 'friend' | 'group'
      calling?: boolean
      participants: Array<ConversationParticipant>
    },
  ) {
    commit('setConversation', payload)
  },
  /**
   * @method setCalling
   * @description Set the calling state of the conversation
   * @example
   * store.dispatch('conversation/setCalling', true);
   */
  setCalling(
    { commit }: ActionsArguments<ConversationState>,
    calling: boolean,
  ) {
    commit('setCalling', calling)
  },
  /**
   * @method addParticipant
   * @description Add a participant to the conversation
   * @example
   * store.dispatch('conversation/addParticipant', participant);
   */
  async addParticipant(
    { commit, state, rootState }: ActionsArguments<ConversationState>,
    address: string,
  ) {
    const friend = rootState.friends.all.find((fr) => fr.address === address)
    const participant = state.participants?.find((p) => p.address === address)

    let peerId = friend?.peerId || participant?.peerId
    if (!peerId) {
      peerId = (
        await createFromPubKey(
          keys.supportedKeys.ed25519.unmarshalEd25519PublicKey(
            new PublicKey(address).toBytes(),
          ).bytes,
        )
      ).toB58String()
    }

    const participantDescriptor = {
      address,
      peerId,
      textilePubkey: friend?.textilePubkey || participant?.textilePubkey,
      name: friend?.name || participant?.name,
      state: friend
        ? friend.state === 'online'
          ? 'CONNECTED'
          : 'DISCONNECTED'
        : participant?.state || 'DISCONNECTED',
      profilePicture: participant?.profilePicture,
      activity: ConversationActivity.NOT_TYPING,
    }

    commit(
      participant ? 'updateParticipant' : 'addParticipant',
      participantDescriptor,
    )
  },
  /**
   * @method addParticipants
   * @description Add multiple participants to the conversation
   * @param participants
   * @example
   * store.dispatch('conversation/addParticipants', participants);
   */
  addParticipants(
    action: ActionsArguments<ConversationState>,
    participants: string[],
  ) {
    participants.forEach((address) => {
      action.dispatch('addParticipant', address)
    })
  },
}
export default actions
