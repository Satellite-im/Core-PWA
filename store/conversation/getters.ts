import {
  ConversationActivity,
  ConversationConnection,
  ConversationParticipant,
  ConversationState,
} from '~/store/conversation/types'
import { RootState } from '~/types/store/store'
import { Friend } from '~/types/ui/friends'
import { Group } from '~/store/groups/types'

const getters = {
  /**
   * @method otherParticipants
   * @description get participants other than yourself
   */
  otherParticipants: (state: ConversationState): ConversationParticipant[] => {
    return state.participants.filter((participant) => participant.name != null)
  },
  /**
   * @method onlineParticipants
   * @description array of online participants other than yourself
   */
  onlineParticipants: (
    state: ConversationState,
    getters: { otherParticipants: ConversationParticipant[] },
  ): ConversationParticipant[] => {
    return getters.otherParticipants.filter(
      (participant) => participant.state === ConversationConnection.CONNECTED,
    )
  },
  /**
   * @method typingParticipants
   * @description array of online participants other than yourself
   */
  typingParticipants: (
    state: ConversationState,
    getters: { otherParticipants: ConversationParticipant[] },
  ): ConversationParticipant[] => {
    return getters.otherParticipants.filter(
      (participant) => participant.activity === ConversationActivity.TYPING,
    )
  },
  /**
   * @method isGroup
   * @description is current recipient group
   */
  isGroup(state: ConversationState) {
    return state.type === 'group'
  },
  /**
   * @method recipient
   * @description current recipient full info
   */
  recipient: (
    state: ConversationState,
    getters: any,
    rootState: RootState,
  ): Friend | Group | undefined => {
    return getters.isGroup
      ? rootState.groups.all.find((group: Group) => group.id === state.id)
      : rootState.friends.all.find(
          (friend: Friend) => friend.peerId === state.id,
        )
  },
}

export default getters
