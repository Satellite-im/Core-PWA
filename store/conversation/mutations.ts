import { ConversationParticipant, ConversationState } from './types'

const mutations = {
  setConversation(
    state: ConversationState,
    payload: {
      id: string
      type: 'friend' | 'group'
      participants: Array<ConversationParticipant>
      calling?: boolean
    },
  ) {
    state.id = payload.id
    state.type = payload.type
    state.calling = !!payload.calling
    state.participants = payload.participants || []
  },
  setCalling(state: ConversationState, active: boolean) {
    state.calling = active
  },
  setParticipants(
    state: ConversationState,
    participants: Array<ConversationParticipant>,
  ) {
    state.participants = participants || []
  },
  addParticipant(
    state: ConversationState,
    participant: ConversationParticipant,
  ) {
    state.participants = [...state.participants, participant]
  },
  updateParticipant(
    state: ConversationState,
    participant: ConversationParticipant,
  ) {
    const prev = state.participants.find(
      (p) =>
        p.address === participant.address ||
        (p.peerId && p.peerId === participant.peerId),
    )
    if (!prev) {
      return
    }
    const next = { ...prev, ...participant }

    // avoid updating the participant unless we have to (video blink issue from peer announce)
    if (JSON.stringify(next) === JSON.stringify(prev)) {
      return
    }

    state.participants = state.participants.map((p) => {
      if (
        p.address === participant.address ||
        (p.peerId && p.peerId === participant.peerId)
      ) {
        return next
      }
      return p
    })
  },
}
export default mutations
