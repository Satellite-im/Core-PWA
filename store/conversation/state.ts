import { ConversationState } from './types'

export default function InitialConversationState(): ConversationState {
  return {
    id: '',
    type: 'friend',
    calling: false,
    participants: [],
  }
}
