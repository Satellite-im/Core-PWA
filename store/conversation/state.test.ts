import * as state from '~/store/conversation/state'

describe('state.default', () => {
  test('0', () => {
    const result: any = state.default()
    expect(result).toMatchSnapshot()
  })
})
