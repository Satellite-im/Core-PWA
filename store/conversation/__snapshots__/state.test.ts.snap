// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`state.default 0 1`] = `
Object {
  "calling": false,
  "id": "",
  "participants": Array [],
  "type": "friend",
}
`;
